lib/support/guest_customization.rb

```diff
@@ -0,0 +1,326 @@
+      windows_timezone_pattern = /^([0-9]+|0x[0-9a-fA-F]+)$/
+
+      input.to_s.match? windows_timezone_pattern
+    end
+
+    # Check for format of Windows Product IDs
+    #
+    # @param [String] input String to check
+    # @returns [Boolean] if value is in Windows Key format
+    def valid_windows_key?(input)
+      windows_key_pattern = /^[A-Z0-9]{5}-[A-Z0-9]{5}-[A-Z0-9]{5}-[A-Z0-9]{5}-[A-Z0-9]{5}$/
+
+      input.to_s.match? windows_key_pattern
+    end
+
+    # Return Guest hostname to be configured and check for validity.
+    #
+    # @returns [String] New hostname to assign
+    def guest_hostname
+      hostname = guest_customization[:hostname] || options[:vm_name]
+
+      hostname_pattern = /^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])$/
+      unless hostname.match?(hostname_pattern)
+        raise Support::GuestCustomizationError.new("Only letters, numbers or hyphens in hostnames allowed")
+      end
+
+      RbVmomi::VIM::CustomizationFixedName.new(name: hostname)
+    end
+
+    # Wait for vSphere task completion and subsequent IP address update (if any).
+    def guest_customization_wait
+      guest_customization_wait_task(guest_customization[:timeout_task] || DEFAULT_TIMEOUT_TASK)
+      guest_customization_wait_ip(guest_customization[:timeout_ip] || DEFAULT_TIMEOUT_IP)
+    end
+
+    # Wait for Guest customization to finish successfully.
+    #
+    # @param [Integer] timeout Timeout in seconds
+    # @param [Integer] sleep_time Time to wait between tries
+    def guest_customization_wait_task(timeout = 600, sleep_time = 10)
+      waited_seconds = 0
+
+      Kitchen.logger.info "Waiting for guest customization (timeout: #{timeout} seconds)..."
+
+      while waited_seconds < timeout
+        events = guest_customization_events
+
+        if events.any? { |event| event.is_a? RbVmomi::VIM::CustomizationSucceeded }
+          return
+        elsif (failed = events.detect { |event| event.is_a? RbVmomi::VIM::CustomizationFailed })
+          # Only matters for Linux, as Windows won't come up at all to report a failure via VMware Tools
+          raise Support::GuestCustomizationError.new("Customization of VM failed: #{failed.fullFormattedMessage}")
+        end
+
+        sleep(sleep_time)
+        waited_seconds += sleep_time
+      end
+
+      raise Support::GuestCustomizationError.new("Customization of VM did not complete within #{timeout} seconds.")
+    end
+
+    # Wait for new IP to be reported, if any.
+    #
+    # @param [Integer] timeout Timeout in seconds. Tools report every 30 seconds, Default: 30 seconds
+    def guest_customization_wait_ip(timeout = 30)
+      return unless guest_customization_ip_change?
+
+      waited_seconds = 0
+
+      Kitchen.logger.info "Waiting for guest customization IP update..."
+
+      while waited_seconds < timeout
+        found_ip = wait_for_ip(timeout, 1.0)
+
+        return if found_ip == guest_customization[:ip_address]
+
+        sleep(sleep_time)
```

Comment:
I encountered an error on this line - undefined local variable sleep_time. Adding `sleep_time =10` as an optional parameter to the method solved the issue.

Reply:
Oops, thanks for the pointer. That's some embarrassing copy+paste error